### DISCUSSION TITLE
[Help] Unable to restore Xcode settings

### DISCUSSION THREAD

**Discussion by bytesizelearning** [mickael-menu/ShadowVim#40] [Help]
### Verify

- [X] I searched the existing discussions for help

### How can we help you?

Trying to run these commands as suggested to restore Xcode settings 

```
defaults write -app Xcode KeyBindingsMode -string 'Default'
defaults delete -app Xcode DVTTextAutoCloseBlockComment
defaults delete -app Xcode DVTTextAutoInsertCloseBrace
defaults delete -app Xcode DVTTextEditorTrimTrailingWhitespace
defaults delete -app Xcode DVTTextEnableTypeOverCompletions
defaults delete -app Xcode DVTTextIndentCaseInC
defaults delete -app Xcode DVTTextUsesSyntaxAwareIndenting
```

I am greeted with the error

```
2023-03-20 22:08:50.356 defaults[37439:1128882] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
2023-03-20 22:08:50.368 defaults[37440:1128887] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
2023-03-20 22:08:50.378 defaults[37441:1128891] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
2023-03-20 22:08:50.386 defaults[37442:1128895] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
2023-03-20 22:08:50.392 defaults[37443:1128899] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
2023-03-20 22:08:50.398 defaults[37444:1128903] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
2023-03-20 22:08:50.404 defaults[37445:1128907] Can't determine domain name for application /Users/carlos/Applications/Xcode.app; defaults unchanged
```

How do I proceed from here? 

Edit:

Solved instead of `Xcode` I needed `/Applications/Xcode.app`

**Comment by mickael-menu** [comment]
As an alternative, I think replacing `-app Xcode` by `com.apple.dt.Xcode` (without a flag) should work too.

**Reply by bytesizelearning** [reply]
I'll try this next time I install, look forward to trying this again when the buffer synchronization has been solved. Thank you and great job this is an amazing project 🥳

**Reply by mickael-menu** [reply]
Thanks, I've been working on using Xcode during Insert mode for improved experience, which means that changing the Xcode settings won't be necessary anymore.

I'm aware of issues with the buffer synchronization and want to tackle these before the next release. But could you describe the issues you encountered to make sure I take them into account?

**Reply by bytesizelearning** [reply]
Sure. One of the things I noticed if you jump to the end of the file it's like you're in insert mode even though you're in normal mode. One of the things I do a lot is `zz` to centre the current line on the screen. There was also the issue when in visual mode the parts that were visually selected weren't highlighted which made it difficult.

I'm not sure if this was user error but I was also unable to get the mapping to work for jump to definition.

**Reply by mickael-menu** [reply]
Thanks for your feedback!

> One of the things I noticed if you jump to the end of the file it's like you're in insert mode even though you're in normal mode.

Ha yeah, this one is annoying. I'm simulating a block caret by selecting a single character in the line. When a line is empty, I select the equivalent of `\n` which displays the whole line as selected, e.g.

<img width="692" alt="Screenshot 2023-03-21 at 11 28 46" src="https://user-images.githubusercontent.com/58686775/226580140-862e877c-6d33-4882-a21b-c534f200cfc2.png">

But in the case of the last empty line, there's no `\n` to select, so I can't show a block caret in normal mode.

I don't have any great solution for this, but what would you think if the `\n` of the previous line was selected instead (only for the EOF situation)? Like this:

<img width="654" alt="Screenshot 2023-03-21 at 11 27 31" src="https://user-images.githubusercontent.com/58686775/226580344-a27dfca6-aa0b-4a65-832b-f1723b5f6606.png">

> One of the things I do a lot is zz to centre the current line on the screen.

I plan to implement it directly in ShadowVim eventually, but there's already a workaround with a custom binding. You can add an Xcode binding with a <kbd>cmd</kbd> modifier for **Center Selection in Visual Area**. Here I chose <kbd>ctrl-alt-cmd-shift-L</kbd>:

<img width="810" alt="Screenshot 2023-03-21 at 11 32 45" src="https://user-images.githubusercontent.com/58686775/226581938-9ff0b362-9082-4f4e-ad89-cc37bd1a4736.png">

Then add in your `init.vim`:

```viml
map zz <Cmd>SVPress <LT>C-M-S-D-l><CR>
```

> There was also the issue when in visual mode the parts that were visually selected weren't highlighted which made it difficult.

Yep, I already implemented this in `develop` and this will be available in the next release.

> I'm not sure if this was user error but I was also unable to get the mapping to work for jump to definition.

Could you share the Xcode binding set for **Jump to Definition** and the `map` command in your `init.vim`?

**Reply by mickael-menu** [reply]
@bytesizelearning I implemented the solution I mentioned for the Normal mode at the end of the buffer. I will use it for a few days to see if it feels natural, so not 100% sure this will make the next release. I would be interested in your opinion too if you want to try it out.

Here's a build: [ShadowVim.app.zip](https://github.com/mickael-menu/ShadowVim/files/11241863/ShadowVim.app.zip)


PR https://github.com/mickael-menu/ShadowVim/pull/48

https://user-images.githubusercontent.com/58686775/232291882-2d2ca7fd-06c4-4c0d-ad94-1b055f0cad1a.mov

**Comment by bytesizelearning** [comment]
You're welcome :)

> But in the case of the last empty line, there's no \n to select, so I can't show a block caret in normal mode.
>
> I don't have any great solution for this, but what would you think if the \n of the previous line was selected instead (only for the EOF > situation)? Like this:

For me personally I like this solution.

> I plan to implement it directly in ShadowVim eventually, but there's already a workaround with a custom binding. You can add an > Xcode binding with a cmd modifier for **Center Selection in Visual Area.** Here I chose `ctrl-alt-cmd-shift-L`:

Awesome it'll be great to have that back because it's the first thing I noticed.

> Could you share the Xcode binding set for **Jump to Definition** and the map command in your `init.vim`?



inside my `init.vim` I tried the mapping both inside and outside the `if` statement

```vim
nmap gd <Cmd>SVPress <LT>C-D-j><CR>
if exists('g:shadowvim')
    " Jump to Definition (⌃⌘J).
    " nmap gd <Cmd>SVPress <LT>C-D-j><CR>
endif
```

I look forward to trying to development release.

**Reply by mickael-menu** [reply]
Your mapping looks good. Make sure you have the default mapping <kbd>⌃⌘J</kbd> in Xcode's Key Bindings settings.

<img width="810" alt="Screenshot 2023-03-21 at 12 40 52" src="https://user-images.githubusercontent.com/58686775/226595826-4d51a719-19a6-4f09-a1a4-ee29d909aa72.png">

**Reply by bytesizelearning** [reply]
Yep that was it! I'd mapped it to something else and totally forgot about it.

Going to have a play around with the window management mappings now because I've missed those. 

Thanks for the help!

Edit: 

I've got marks again 🥳
